exp/metadataserver/server.go

```diff
@@ -0,0 +1,173 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package metadataserver
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+
+	enterrors "github.com/weaviate/weaviate/entities/errors"
+	"github.com/weaviate/weaviate/exp/metadataserver/proto/api"
+
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	grpc_sentry "github.com/johnbellone/grpc-middleware-sentry"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+type Server struct {
+	grpcServer                *grpc.Server
+	listenAddress             string
+	grpcMessageMaxSize        int
+	sentryEnabled             bool
+	querierManager            *QuerierManager
+	dataEventsChannelCapacity int
+	log                       *logrus.Logger
+}
+
+func NewServer(listenAddress string, grpcMessageMaxSize int,
+	sentryEnabled bool, querierManager *QuerierManager, dataEventsChannelCapacity int, log *logrus.Logger,
+) *Server {
+	return &Server{
+		listenAddress:             listenAddress,
+		grpcMessageMaxSize:        grpcMessageMaxSize,
+		sentryEnabled:             sentryEnabled,
+		querierManager:            querierManager,
+		dataEventsChannelCapacity: dataEventsChannelCapacity,
+		log:                       log,
+	}
+}
+
+// Open starts the server and registers it as the cluster service server. Blocking.
+// Returns an error if the configured listenAddress is invalid.
+// Returns an error if the configured listenAddress is un-usable to listen on.
+func (s *Server) Open() error {
```

Comment:
Suggestion: I like long running blocking api like this to accept `ctx` and not expose separate `Close()` api.  Should respect the context cancellation and cleanup the resources if it's cancelled.

Something like

```go
svr = metadata.NewServer()
ctx, cancel = context.WithCancel()
defer cancel()

svr.Serve(ctx)
```

Rationale
1. Note. I don't need any extra go routines to start and stop. Because start() is blocking, you need another goroutine to stop it, 
2. Similar to http.ServeAndListen - Also blocking long-running api that just returns error

wdyt?

Reply:
Thanks, for the suggestion, I've added this. Let me know if you have feedback on the implementation, especially if you can think of any issues with how I'm using context's as it's easy to make subtle mistakes with them